firedrake/adjoint_utils/blocks/solving.py

```diff
@@ -700,8 +723,10 @@ def _ad_assign_map(self, form, solver):
                            firedrake.Cofunction)):
                 coeff_count = coeff.count()
                 if coeff_count in form_ad_count_map:
-                    assign_map[form_ad_count_map[coeff_count]] = \
-                        block_variable.saved_output
+                    if solver == Solver.HESSIAN:
+                        assign_map[form_ad_count_map[coeff_count]] = block_variable.tlm_value
+                    else:
+                        assign_map[form_ad_count_map[coeff_count]] = block_variable.saved_output
 
         if (
             solver == Solver.ADJOINT
```

Comment:
Will this make sure that the state is updated correctly for the Hessian solve?
```suggestion
            (solver in (Solver.ADJOINT, Solver.HESSIAN))
```